icu4c/source/i18n/unitconverter.cpp

```diff
@@ -49,10 +50,24 @@ MeasureUnit extractBaseUnit(const MeasureUnit &source,
         // Multiply the power of the singleUnit by the power of the baseUnit. For example, square-hectare
         // must be p4-meter. (NOTE: hectare --> square-meter)
         auto baseUnit = MeasureUnit::forIdentifier(rateInfo.baseUnit.toStringPiece(), status);
-        auto singleBaseUnit = SingleUnitImpl::forMeasureUnit(baseUnit, status);
-        singleBaseUnit.dimensionality *= singleUnit.getDimensionality(status);
-
-        result = result.product(singleBaseUnit.build(status), status);
+        if (dimensionality != 1) {
+            if (baseUnit.getComplexity(status) == UMEASURE_UNIT_SINGLE) {
+                baseUnit = baseUnit.withDimensionality(
+                    dimensionality * baseUnit.getDimensionality(status), status);
+            } else {
+                // Single units like volt, joule, newton, and knots map to
+                // compound base units. Fortunately square-volt, square-joule,
+                // square-newton, and square-knots don't generally make sense.
+                // We treat them as illegal.
+                status = U_ILLEGAL_ARGUMENT_ERROR;
+                return MeasureUnit();
+                // TODO(younies,hugovdm): (dimensionality = -1) might appear
+                // though? one-per-joule, one-per-volt? Consider this. Find some
+                // reasonable real-world usages that would trigger this problem.
```

Comment:
I'll bring this up with Mark as well to discuss the specification for unit identifiers, since the spec doesn't (currently) protect us from such weird units. I want to figure out how to best document our choices (and what choices to make).